### ISSUE TITLE
feat: Implement a java.util.logging.Handler which produces the google-fluentd JSON format

### ISSUE THREAD

**Issue by bpcreech** [googleapis/java-logging#746]
**Is your feature request related to a problem? Please describe.**

The Cloud Run team advises that, on that platform, and App Engine gen 2, it's best to use the [google-fluentd JSON format](https://cloud.google.com/logging/docs/agent/logging/configuration#special-fields), writing JSON log records to stderr.

However, there's no `java.util.logging.Handler` which produces that format.

**Describe the solution you'd like**

Can we get a sibling class of `com.google.cloud.logging.LoggingHandler` which produces google-fluentd-style logs to stderr?

The format is *almost* the same as the `LogEntry` proto, except with some fields renamed, and others dropped, so similar code could do the conversion (in fact, I've already written this code, albeit not in a totally general way).

One could potentially refactor `LoggingHandler` slightly to get at the converted record. However, the `logEntryFor` method needs to be made public, or better, logic needs to be refactored (e.g., by creating a separate `LogRecordConverter` class which converts from `java.util.logging.LogRecord` to `com.google.cloud.logging.LogEntry`).

**Describe alternatives you've considered**

One could of course simply use `LoggingHandler`, but the advice from GCP Serverless is to use the JSON format. And it's a fairly ubiquitous format (used by both the Serverless platform and the fluentd plugin) so it seems like it would be a good idea for this library to support it.

**Additional context**

Contact me for an already-existing internal implementation which would need to be generalized somewhat for inclusion here. :)

**Comment by minherz** [comment]
hi, would you mind to provide a link to the documentation that recommends writing google-fluentd formatted log records to STDERR?
Google Cloud Serverless platforms are deployed with logging agents that captures output to STDOUT and STDERR. If the output is formatted as Json the agents try to parse it for special fields which then interpreted according to the schema that you had [referenced](https://cloud.google.com/logging/docs/agent/logging/configuration#special-fields).
To have it working you do not need to use the client logging library. You can just print to console a data in the Json format.

We work on the feature that enrich currently captured logs with additional metadata (e.g. source location for logs in debug severity, etc.). For Java, the [logback](https://github.com/googleapis/java-logging-logback) library will have an adapter that will be designed to leverage the logging agent on the Serverless platforms. Similar functionality will be added to JUL as well.

**Comment by bpcreech** [comment]
Hi! Right, [this](https://cloud.google.com/logging/docs/agent/logging/configuration#special-fields) is the doc describing the google-fluentd format, and [this](https://cloud.google.com/run/docs/logging#writing_structured_logs) is the doc recommending use of that format from Serverless (well in that case specifically Cloud Run Managed).

What I'm looking for is a `java.util.logging.Handler` which ***produces** that JSON format*. That would enable anything using ordinary Java logging to write the structured log format without any extra code.

I think this is the exactly feature you allude to when you said *"the feature that enrich currently captured logs with additional metadata...Similar functionality will be added to JUL as well"*. Is there any issue currently open for that?

**Comment by minherz** [comment]
The referenced documentation does not define the use of client library as a recommended practice. If you read the [document](https://cloud.google.com/run/docs/logging#container-logs) it states **4** different methods to do the same thing.

I understand that you use JUL to write logs. Currently the LoggingHandler for JUL captures severity and resource metadata beside the message payload. I think we discussed in another issue your request to enhance the functionality as you see it fit.
What data are you looking to capture as a structured log via JUL?
It would be helpful to see an example of the log entry that you see in Log Explorer and cannot get using LoggingHandler.

**Comment by steren** [comment]
(Cloud Run PM here) 

Capturing logs via `stderr` or `stdout` is considered slightly more robust than calling the Cloud Logging API.

Unfortunately, Google Cloud doesn't provide helper client libraries that are able to send JSON logs to `stderr` or `stdout`, they only send to the Cloud Logging API.

**Comment by minherz** [comment]
thank you @steren. Correct me if I am wrong, the agent in Cloud Run populates `resource` metadata for each line captured from `stderr` or `stdout`. And additionally it interpretes the Json formatted lines by trying to find additional fields ([by name](https://cloud.google.com/logging/docs/agent/logging/configuration#special-fields)).

there is a work-in-progress (see #691) that implements auto-population of LogEntry metadata implicitly. We plan to integrate this behavior with JUL LoggingHandler and logback adapter.

however, this behavior does not aim to change the way how JUL handler captures information from LogRecord which is how I read this FR. i hope that after @bpcreech provides an example that shows what data is expected to be published which is not published now, we will have a clear picture about this FR and will be able to provide a relevant feedback.

**Comment by bpcreech** [comment]
For me: I want to publish:

1. severity
2. message
3. time
4. labels (to capture various `java.util.logging.LogRecord` fields like thread ID which have no home in `LogEntry`)
5. sourceLocation
6. spanId
7. trace
8. traceSampled

This is what our internal JSON converter already does and what I'm trying to get generalized here.

**Comment by minherz** [comment]
I would say that this FR partially overlaps with #747. Given there is a good way to inherit the current conversion from `LogRecord` to `LogEntry` and extend it according to in-house requirements, it should be easy comparing to the current implementation.
On this account, printing to `stdout` instead of calling write API can be done if preferred.

@bpcreech if you agree we can close this one as a duplication of #747.
As you suggested the quick fix will be to change the method `logEntryFor` visibility to be `protected` instead of `private`.

**Comment by minherz** [comment]
Duplicated by #800

**Comment by bpcreech** [comment]
Sorry for much later response: Yeah this was intended to be blocked by #747. #747 is a probable implementation method. The fluentd-style JSON handler first converts the JUL `LogRecord` to the Cloud Logs proto by just calling `logEntryFor`, then (via whatever mechanism) writes it out as JSON.

And yeah I guess this issue is the same as #800.
